### PR TITLE
chore: Refactor multi-domain communication lifecycle

### PR DESCRIPTION
**Pull request by chrisbreiding**
Refactors the communication lifecycle and how command and logs are handled for multi-domain.

Previously, commands in the secondary domain had proxy versions in the primary domain. A similar mechanism was in place for logs. This meant that they had to be kept in sync and that certain event listeners had to set up and torn down at the right time in order to support them. This led to various timing issues and complicated workarounds. It also required adding a convoluted pausing mechanism to the command queue in the secondary.

Since the command proxies only needed to exist in the primary domain for the purpose of sending messages to the reporter, it's easier to do that directly (more or less) from the the actual command instances in the secondary. The same applies for logs. This simplifies things greatly as the events are simply a stream and can be listened to globally, removing the need to worry about setting them up and tearing them down in sync with the switchToDomain command.

One unfortunate side effect is that this PR removes support for the `done` callback. It was failing with the refactor, and since we had decided not to support it inside the switchToDomain callback after all, I didn't deem it worthwhile to spend time fixing it. As demonstrated in this PR, promises can be used instead of done callbacks.

This PR also makes some file name and test name changes that aren't strictly necessary, they're just some cleanup I did while implementing the refactor.

### PR Tasks

- [x] Have tests been added/updated?
- N/A Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- N/A Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- N/A Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- N/A Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/driver/cypress/integration/e2e/multi-domain/navigation_spec.ts
Comment (chrisbreiding) [inline packages/driver/cypress/integration/e2e/multi-domain/navigation_spec.ts:103]: Logs in the secondary no longer have a counterpart in the primary, so these tests needed to change accordingly.

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_yield_spec.ts
Comment (chrisbreiding) [inline packages/driver/cypress/integration/e2e/multi-domain/multi_domain_yield_spec.ts:125]: This fails quicker than `invoke('text')` because `invoke` will retry until the timeout (4000ms) whereas `then` does not retry.

File: packages/driver/cypress/integration/e2e/multi-domain/commands/multi_domain_misc.spec.ts
Comment (chrisbreiding) [inline packages/driver/cypress/integration/e2e/multi-domain/commands/multi_domain_misc.spec.ts:80]: Example of using a promise instead of a done callback.

File: packages/driver/cypress/integration/e2e/multi-domain/commands/multi_domain_misc.spec.ts
Comment (mjhenkes) [inline packages/driver/cypress/integration/e2e/multi-domain/commands/multi_domain_misc.spec.ts:36]: much nicer, my previous implementation didn't print out the diff like i expected it to

File: packages/runner-shared/src/event-manager.js
Comment (mschile) [inline packages/runner-shared/src/event-manager.js:129]: When is Cypress not set? Maybe add a comment.

**Review by mschile — APPROVED** [review]
Looks good! Glad to see all the complexity removed!

File: packages/driver/src/cy/multi-domain/unserializable_subject_proxy.ts
Comment (mjhenkes) [inline packages/driver/src/cy/multi-domain/unserializable_subject_proxy.ts:40]: i wish our linter complained about un-used arguments

Comment (chrisbreiding) [reply packages/driver/src/cy/multi-domain/unserializable_subject_proxy.ts:40]: I agree. We used to use the no-unused-vars rule (or something similar that applies to this situation) but it was removed at some point and I'm not sure why.

File: packages/driver/src/multi-domain/communicator.ts
Comment (mjhenkes) [inline packages/driver/src/multi-domain/communicator.ts:11]: This function isn't actually serializing, the actual serializing is done by the postMessage. This function massages the object into a shape that can be serialized.

File: packages/driver/src/multi-domain/communicator.ts
Comment (mjhenkes) [inline packages/driver/src/multi-domain/communicator.ts:183]: this and to with error below feel messy to me. There isn't any contract between the event and the data, any event could send a subject but would be limited to just the subject. Should we switch to get rid of these special events and instead treat the err and subject keys special in the toPrimary command itself? It might be nice to have a try catch with an error just in case something can't be serialized to at least send an error for an event.

File: packages/driver/src/multi-domain/domain_fn.ts
Comment (mjhenkes) [inline packages/driver/src/multi-domain/domain_fn.ts:36]: I'm curious what would happen if we synced stability with the main domain, but thats an experiment for another day.

File: packages/driver/src/multi-domain/domain_fn.ts
Comment (mjhenkes) [inline packages/driver/src/multi-domain/domain_fn.ts:61]: what is the nestedIndex?

File: packages/driver/src/multi-domain/communicator.ts
Comment (mjhenkes) [reply packages/driver/src/multi-domain/communicator.ts:183]: I could make changes around this in my multi switchToDomains PR if you want to defer it to later.

File: packages/driver/cypress/integration/e2e/multi-domain/commands/multi_domain_actions.spec.ts
Comment (mschile) [reply packages/driver/cypress/integration/e2e/multi-domain/commands/multi_domain_actions.spec.ts]: I saw that you made the change here but this comment really applies to multiple tests.

File: packages/driver/src/cy/multi-domain/index.ts
Comment (AtofStryker) [inline packages/driver/src/cy/multi-domain/index.ts:75]: Did we want to call `correctStackForCrossDomainError` here or save that for a follow up or #20256? Some of the returned errors are bit lengthy 😅

File: packages/driver/src/multi-domain/communicator.ts
Comment (chrisbreiding) [reply packages/driver/src/multi-domain/communicator.ts:183]: > Should we switch to get rid of these special events and instead treat the err and subject keys special in the toPrimary command itself? It might be nice to have a try catch with an error just in case something can't be serialized to at least send an error for an event.

I like that idea.

**Review by AtofStryker — CHANGES_REQUESTED** [review]

File: packages/driver/src/multi-domain/domain_fn.ts
Comment (chrisbreiding) [reply packages/driver/src/multi-domain/domain_fn.ts:46]: This code is cribbed from the primary domain `cy.ts`'s [doneEarly](https://github.com/cypress-io/cypress/blob/75c87502a2ca88fa76c8c5bb5f52626e0f70b08f/packages/driver/src/cypress/cy.ts#L1226) and [cleanup](https://github.com/cypress-io/cypress/blob/75c87502a2ca88fa76c8c5bb5f52626e0f70b08f/packages/driver/src/cypress/cy.ts#L1052) methods. It was originally pulled out when we implemented done callback support, as it didn't seem to work to call the cy `doneEarly` method directly, I'm going to see if it's possible to do that now.

File: packages/driver/cypress/integration/e2e/multi-domain/navigation_spec.ts
Comment (mjhenkes) [reply packages/driver/cypress/integration/e2e/multi-domain/navigation_spec.ts:75]: this is another one i wish we had linters around

File: packages/driver/src/multi-domain/communicator.ts
Comment (chrisbreiding) [reply packages/driver/src/multi-domain/communicator.ts]: Addressed in https://github.com/cypress-io/cypress/pull/20247/commits/d40791761dc9018b5383a2ee6ed6862c80ddb82c

File: packages/driver/src/multi-domain/domain_fn.ts
Comment (chrisbreiding) [reply packages/driver/src/multi-domain/domain_fn.ts]: Addressed in https://github.com/cypress-io/cypress/pull/20247/commits/9bc9fcb931e3e09954ba6f81d395f950fe697098

File: packages/driver/src/multi-domain/domain_fn.ts
Comment (chrisbreiding) [reply packages/driver/src/multi-domain/domain_fn.ts]: No longer relevant after https://github.com/cypress-io/cypress/pull/20247/commits/9bc9fcb931e3e09954ba6f81d395f950fe697098

File: packages/driver/src/multi-domain/domain_fn.ts
Comment (chrisbreiding) [reply packages/driver/src/multi-domain/domain_fn.ts]: See https://github.com/cypress-io/cypress/blob/75c87502a2ca88fa76c8c5bb5f52626e0f70b08f/packages/driver/src/cypress/cy.ts#L1141

File: packages/driver/cypress/integration/e2e/multi-domain/commands/multi_domain_misc.spec.ts
Comment (chrisbreiding) [reply packages/driver/cypress/integration/e2e/multi-domain/commands/multi_domain_misc.spec.ts]: Done in https://github.com/cypress-io/cypress/pull/20247/commits/170826a55135be9915a0baaf44988c76bc52ccd0

File: packages/driver/cypress/integration/e2e/multi-domain/commands/multi_domain_misc.spec.ts
Comment (chrisbreiding) [reply packages/driver/cypress/integration/e2e/multi-domain/commands/multi_domain_misc.spec.ts]: Done in https://github.com/cypress-io/cypress/pull/20247/commits/170826a55135be9915a0baaf44988c76bc52ccd0

File: packages/driver/cypress/integration/e2e/multi-domain/commands/multi_domain_misc.spec.ts
Comment (chrisbreiding) [reply packages/driver/cypress/integration/e2e/multi-domain/commands/multi_domain_misc.spec.ts]: Yes - your comments made me realize that this assertion would never have failed because at this point `addedCommands` would have to be empty to not fail the check above. I've refactored it now so it should catch every scenario. 

Done in https://github.com/cypress-io/cypress/pull/20247/commits/170826a55135be9915a0baaf44988c76bc52ccd0

File: packages/driver/src/cy/multi-domain/util.ts
Comment (chrisbreiding) [reply packages/driver/src/cy/multi-domain/util.ts:10]: Yeah, I'll revert this and leave it up to your PR.

Comment (chrisbreiding) [reply packages/driver/src/cy/multi-domain/util.ts]: Done in https://github.com/cypress-io/cypress/pull/20247/commits/7f8ab7e333ee2f4dabd2299f5baa4084aa9baafa

File: packages/driver/src/cy/multi-domain/index.ts
Comment (chrisbreiding) [reply packages/driver/src/cy/multi-domain/index.ts]: Done in https://github.com/cypress-io/cypress/pull/20247/commits/3c81efaa48edf272cd1aaa80e12a779c95684397

File: packages/driver/src/cy/multi-domain/index.ts
Comment (chrisbreiding) [reply packages/driver/src/cy/multi-domain/index.ts:75]: Looks like that only happens if you leave the done in the test after removing the fail handler, which isn't really a valid test. It's a much cleaner error if you remove the done. I'm not sure why it isn't showing the code frame though...

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_spec.ts
Comment (chrisbreiding) [reply packages/driver/cypress/integration/e2e/multi-domain/multi_domain_spec.ts:331]: Probably would be a good idea to add cy-in-cy tests once we're merged with the 10.0 tests. I removed these tests since they're related to done callbacks.

File: packages/driver/src/cy/multi-domain/index.ts
Comment (AtofStryker) [reply packages/driver/src/cy/multi-domain/index.ts:75]: Makes sense. I can just picture a user running into something like this and having some issues parsing what went wrong initially. I think the codeFrame is missing since we aren't reprocessing the error coming back from the secondary, but not 100% sure.

**Review by AtofStryker — APPROVED** [review]
🎉

**Pull request merged**
